src/db.rs

```diff
@@ -203,29 +185,150 @@ impl DbInner {
 	}
 
 	fn get(&self, col: ColId, key: &[u8]) -> Result<Option<Value>> {
-		let key = self.columns[col as usize].hash(key);
-		let overlay = self.commit_overlay.read();
-		// Check commit overlay first
-		if let Some(v) = overlay.get(col as usize).and_then(|o| o.get(&key).map(|(_, v)| v.clone())) {
-			return Ok(v);
+		match &self.columns[col as usize] {
+			Column::Hash(column) => {
+				let key = column.hash(key);
+				let overlay = self.commit_overlay.read();
+				// Check commit overlay first
+				if let Some(v) = overlay.get(col as usize).and_then(|o| o.get(&key)) {
+					return Ok(v);
+				}
+				// Go into tables and log overlay.
+				let log = self.log.overlays();
+				let key = TableKey::Partial(key);
+				column.get(&key, log)
+			},
+			Column::Tree(column) => {
+				let overlay = self.commit_overlay.read();
+				if let Some(l) = overlay.get(col as usize).and_then(|o| o.btree_get(key)) {
+					return Ok(l.cloned());
+				}
+				// We lock log as btree structure changed while reading it would be an issue.
+				let log = self.log.overlays().read();
+				column.with_locked(|btree| BTreeTable::get(key, &*log, btree))
+			},
 		}
-		// Go into tables and log overlay.
-		let log = self.log.overlays();
-		self.columns[col as usize].get(&key, log)
 	}
 
 	fn get_size(&self, col: ColId, key: &[u8]) -> Result<Option<u32>> {
-		let key = self.columns[col as usize].hash(key);
-		let overlay = self.commit_overlay.read();
-		// Check commit overlay first
-		if let Some(l) = overlay.get(col as usize).and_then(
-			|o| o.get(&key).map(|(_, v)| v.as_ref().map(|v| v.len() as u32))
-		) {
-			return Ok(l);
+		match &self.columns[col as usize] {
+			Column::Hash(column) => {
+				let key = column.hash(key);
+				let overlay = self.commit_overlay.read();
+				// Check commit overlay first
+				if let Some(l) = overlay.get(col as usize).and_then(|o| o.get_size(&key)) {
+					return Ok(l);
+				}
+				// Go into tables and log overlay.
+				let log = self.log.overlays();
+				let key = TableKey::Partial(key);
+				column.get_size(&key, log)
+			},
+			Column::Tree(_column) => unimplemented!("TODO any use over simply using get"),
```

Comment:
Maybe just add a dummy impl here that calls self.get() and maps result to a size